Added nseries for bessel functions

0sidharth opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Continuing https://github.com/sympy/sympy/pull/17303

The original PR has many issues and requires a lot of fixes, so I am breaking it up into 3 different PRs, 1 for each special function.

#### Brief description of what is fixed or changed
Added `nseries` expansion for bessel function
```py
In [2]: besselj(1, x).series(x)
Out[2]: 
     3     5        
x   x     x     ⎛ 6⎞
─ - ── + ─── + O⎝x ⎠
2   16   384        

```

#### Other comments
There is an issue with `besselj(n, x).series(x)` that returns `NotImplementedError: Result depends on the sign of sign(n - 6)`. This issue is arising due to the Order term which is x^n O(x**6) and I think sympy does not know how to handle this for any integer `n <= -1`. I can't see how it would be possible to resolve this, so `besselj(n, x).series(x)` is not supported currently.

The commit from the Original PR has not been carried forward as there was only 1 commit and adding that commit would mean having to add another commit which removes some of the changes added in the previous commit, which I don't think is good. Due credits to @ArighnaIITG 

For `bessely` expansion, I have used [this](https://functions.wolfram.com/Bessel-TypeFunctions/BesselY/06/01/04/01/02/0008/)
For `besselj` expansion, I have used [this](https://functions.wolfram.com/Bessel-TypeFunctions/BesselJ/06/01/04/01/01/0003/)

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* functions
    * Added `nseries` expansion for besselj and bessely 
<!-- END RELEASE NOTES -->

0sidharth committed:
Added power series expansion for bessel functions

0sidharth commented:
Is the test failure related to the changes, or is it a bug/random failure? It is passing locally
```
______________ sympy/utilities/tests/test_timeutils.py:test_timed ______________
Traceback (most recent call last):
  File "/home/runner/work/sympy/sympy/sympy/utilities/tests/test_timeutils.py", line 10, in test_timed
    assert result[0] == 100000 and result[3] == "ns"
AssertionError
```

0sidharth commented:
Closed by mistake

jksuom commented on sympy/functions/special/tests/test_bessel.py:
```diff
@@ -33,6 +33,11 @@ def test_bessel_twoinputs():
+def test_bessel_series():
+    from sympy import EulerGamma
+    assert besselj(0, x).series(x) == 1 - x**2/4 + x**4/64 + O(x**6)
+    assert bessely(0, x).series(x) == (2*(log(x/2) + EulerGamma)*besselj(0, x) + x**2/2 - 3*x**4/64 + O(x**6))/pi
```
Maybe `besselj` should also be expanded.

jksuom commented on sympy/functions/special/tests/test_bessel.py:
```diff
@@ -33,6 +33,12 @@ def test_bessel_twoinputs():
+    assert besselj(0, x).series(x) == 1 - x**2/4 + x**4/64 + O(x**6)
+    assert bessely(0, x).series(x) == (2*S.EulerGamma - 2*log(2) + 2*log(x) \
+        + x**2/2 + x**2*log(2)/2 - S.EulerGamma*x**2/2 - x**2*log(x)/2 \
+        + S.EulerGamma*x**4/32 - x**4*log(2)/32 - 3*x**4/64 + x**4*log(x)/32 + O(x**6))/pi
```
The order term should not be in the parentheses as this is not an asymptotic expansion. Besides, it looks like it should be `O(x**6*log(x))` which dominates `O(x**6)`.

0sidharth replied:
Oh, right. For some reason, I had thought the `log(x)` was not needed, but that is incorrect.

jksuom commented on sympy/functions/special/bessel.py:
```diff
@@ -298,11 +308,32 @@ def _eval_rewrite_as_besseli(self, nu, z, **kwargs):
+        if z.limit(x, 0) is nu is S.Zero:
+            k = Symbol('k')
+            term = (-1)**(k + 1) * harmonic(k) * (z / 2)**(2*k) / (factorial(k)**2)
+            s = [term.subs(k, j) for j in range(0, n) if 2 * j < n]
```
The term expression could be included in the list comprehension with something like `for k in range((n+1)//2)`. A new symbol need not be created.

0sidharth replied:
OK I will do that. In the original PR, it was mentioned to break up the term calculation like this to make it easier to debug.

jksuom replied:
I agree that a fixed range together with a condition like `2 * j < n` is more readable than a computed range, but it seems that a computation is needed in general. For example, if `z = cbrt(x)` then a range of ca. 3n/2 terms is necessary.

0sidharth replied:
`z = cbrt(x)` or even `sqrt(x)` was leading to infinite recursion due to besselj series not being implemented. It is now implemented and fixed

jksuom commented on sympy/functions/special/bessel.py:
```diff
@@ -216,11 +217,30 @@ def _eval_rewrite_as_bessely(self, nu, z, **kwargs):
+
+        if z.limit(x, 0) is S.Zero:
+            s = [(-1)**k * z**(2 * k) / (factorial(k) * factorial(nu + k)
+                    * 2**(2 * k)) for k in range((n + 1)//2)] + [Order(z**(n))]
```
More terms are needed if the order of z in x is less than 1, eg.  sqrt(x), cbrt(x)... (The range can be computed from the leading term of z.) I would also add the last Order term only at the return line so it would be `Order(x**n)` which may be different from `Order(z**n)`.

0sidharth replied:
Currently - 
```
In [8]: besselj(0, sqrt(x)).series(x)
Out[8]: 
         2     3       4         5           
    x   x     x       x         x        ⎛ 6⎞
1 - ─ + ── - ──── + ────── - ──────── + O⎝x ⎠
    4   64   2304   147456   14745600        

In [9]: besselj(1, sqrt(x)).series(x)
Out[9]: 
      3/2    5/2     7/2      9/2       11/2          
√x   x      x       x        x         x          ⎛ 6⎞
── - ──── + ──── - ───── + ─────── - ───────── + O⎝x ⎠
2     16    384    18432   1474560   176947200        

```
It seems like there are sufficient terms and I think the caller is taking care of n?

jksuom replied:
There are sufficiently many terms because `_eval_nseries` is called repeatedly from `expr.series` with increasing values of `n`:  https://github.com/sympy/sympy/blob/b564d9ba2705ee2978766e1ee2102750834df68b/sympy/core/expr.py#L3002-L3016
However, that is inefficient and we would want to have a sufficient number of terms on the first call. (This whole piece of code looks like a hack that was created to mask errors in various `_eval_nseries` methods. Besides, it will add at most 9 terms which is not enough in general.)

0sidharth replied:
Oh, I hadn't noticed that. Then perhaps, the block of code you pointed out should be removed in the future, by making appropriate corrections to the `_eval_nseries` methods of functions that need it.

0sidharth replied:
Only two tests in the `series` module fail on commenting this block - 
```py
assert D(exp(x), x).series(n=3) == D(1, x) + D(x, x) + D(x**2/2, x) + D(x**3/6, x) + O(x**3) # where D = Derivative
assert Derivative(sin(x), x).series(x, n=3).doit() == cos(x).series(x, n=3)
```
will look into this

0sidharth committed:
bessel series: Added computation of number of terms in case of power < 1

jksuom commented on sympy/functions/special/bessel.py:
```diff
@@ -234,10 +234,15 @@ def _eval_nseries(self, x, n, logx, cdir=0):
 
+        # In case of powers less than 1, number of terms need to be computed
+        # separately to avoid repeated callings of _eval_nseries with wrong n
+        _, exp = z.as_leading_term(x).as_base_exp()
```
Will this work also if the leading term is something like `2*sqrt(x)`?

0sidharth replied:
It does not work, `as_base_exp()` returns `(2*sqrt(x), 1)` in this case. I will try to come up with a fix/alternative

jksuom commented on sympy/functions/special/bessel.py:
```diff
@@ -234,10 +234,15 @@ def _eval_nseries(self, x, n, logx, cdir=0):
+        # In case of powers less than 1, number of terms need to be computed
+        # separately to avoid repeated callings of _eval_nseries with wrong n
+        _, exp = z.as_leading_term(x).as_base_exp()
+        newn = n//exp + 1
```
Does this work also when `exp` is not an integer?

jksuom replied:
I would try something like `ceiling(n/exp)`.

0sidharth replied:
Yes, I believe it is working for float `exp`. Here is the output (with `print(exp, newn)`) - 
```py
In [3]: besselj(1, x).series(x)
1 7
Out[3]: 
     3     5        
x   x     x     ⎛ 6⎞
─ - ── + ─── + O⎝x ⎠
2   16   384        

In [4]: besselj(1, x**(2.5)).series(x)
2.50000000000000 3
Out[4]: 
 2.5        
x       ⎛ 6⎞
──── + O⎝x ⎠
 2          

In [5]: besselj(1, x**(0.5)).series(x)
0.500000000000000 13
Out[5]: 
      5.5        4.5      3.5    2.5    1.5    0.5        
     x          x        x      x      x      x       ⎛ 6⎞
- ───────── + ─────── - ───── + ──── - ──── + ──── + O⎝x ⎠
  176947200   1474560   18432   384     16     2          
```

jksuom replied:
newn should also be an integer.

0sidharth replied:
But newn is indeed an integer. `n//exp` is floor division and I am adding an integer (1) to it. I think ceiling(n/exp) is probably a better way to write this line, so will make the change

0sidharth commented on sympy/functions/special/bessel.py:
```diff
@@ -216,11 +218,35 @@ def _eval_rewrite_as_bessely(self, nu, z, **kwargs):
+
+        # In case of powers less than 1, number of terms need to be computed
+        # separately to avoid repeated callings of _eval_nseries with wrong n
+        exp = max((s.as_base_exp()[1] for s in z.atoms(Pow)), default=1)
```
I considered using `_, exp = LM(z).as_base_exp()`, but could not see `LM` function being used anywhere else in sympy library code, so should it not be used?

0sidharth commented on sympy/functions/special/bessel.py:
```diff
@@ -216,11 +218,35 @@ def _eval_rewrite_as_bessely(self, nu, z, **kwargs):
+
+        # In case of powers less than 1, number of terms need to be computed
+        # separately to avoid repeated callings of _eval_nseries with wrong n
+        exp = max((s.as_base_exp()[1] for s in z.atoms(Pow) if s.has(x)), default=1)
```
I considered using `_, exp = LM(z).as_base_exp()`, but could not see `LM` function being used anywhere else in sympy library code, so should it not be used?

Also, for `z = exp(x**(-3)) + x**(-4)`, this returns `-3` but `as_leading_term(x)` also returns `exp(x**(-3))`

0sidharth replied:
I found some places in the integrals module which use a similar function `LC`. So is it alright if I use `LM` here?

jksuom replied:
`leadterm` may be the solution. It is a bit inconvenient to use though as it may raise and has to be put in a try-except block.

0sidharth replied:
I will squash unnecessary commits after the code is accepted

jksuom commented on sympy/functions/special/bessel.py:
```diff
@@ -216,11 +218,38 @@ def _eval_rewrite_as_bessely(self, nu, z, **kwargs):
+        # separately to avoid repeated callings of _eval_nseries with wrong n
+        try:
+            _, exp = z.leadterm(x)
+            newn = ceiling(n/exp) if not exp is S.Zero else n
```
`exp` could also be negative. I would leave the above `leadterm` line alone in the try block and replace the condition `z.limit(x, 0) is S.Zero` by `exp > 0` (and only then compute `newn`).

jksuom commented:
Can you add some tests with arguments of various orders? Maybe also something like `sqrt(x) + x`.

0sidharth committed:
Added bessel series tests

jksuom commented:
These examples look good. Now there could be some more general examples where `z` is something like `x/(1 - x)` or `sin(x)`.

0sidharth commented:
I was having some doubts regarding the `bessely` expansion the original author implemented, and anyways that was for the specific case of `nu == 0`. I have modified the series expansion to reflect the one found on [wolfram](https://functions.wolfram.com/Bessel-TypeFunctions/BesselY/06/01/04/01/02/) and other places on the internet, and also verified the results with wolfram alpha. I have also added more generalized tests.

jksuom commented on sympy/functions/special/bessel.py:
```diff
@@ -216,11 +217,40 @@ def _eval_rewrite_as_bessely(self, nu, z, **kwargs):
+            s = [(-1)**k * z**(2 * k) / (factorial(k) * factorial(nu + k)
+                    * 2**(2 * k)) for k in range((newn + 1) // 2)]
+            r = [k._eval_nseries(x, n, logx, cdir) for k in s]
+            t = z._eval_nseries(x, n, logx, cdir)
```
`z._eval_nseries` will be called repeatedly on the line above (for each `z**(2*k)`). Would it be more efficient to compute it once and use the result (without the order term) in `s`?

0sidharth replied:
I see my mistake, the method you suggested would definitely be much faster.

jksuom replied:
Maybe we should first compute the expansion of `z**2`, and then iteratively `z**(2*k + 2) = z**(2*k)*z**2`.

0sidharth replied:
I didn't fully understand what you meant. Are you referring to something like this?
```py
    newn = ceiling(n/exp)                                                       
    r = z._eval_nseries(x, n, logx, cdir).removeO()                             
    t = (z**2)._eval_nseries(x, n, logx, cdir).removeO()                        
    s = []                                                                      
    p = 1                                                                       
    for k in range((newn + 1)//2):                                              
        term = (-1)**k * p / (factorial(k) * factorial(nu + k)* 2**(2 * k))     
        s.append(term)                                                          
        p *= t                                                                  
    return ((r / 2)**nu * Add(*s)).expand() + Order(x**n, x)                    
```

jksuom replied:
Yes, something like that.

0sidharth committed:
Implemented recursive computation of bessel series

jksuom commented on sympy/functions/special/bessel.py:
```diff
@@ -241,12 +241,17 @@ def _eval_nseries(self, x, n, logx, cdir=0):
             return self
 
-        if exp.is_positive:
+        if exp.is_positive and nu.is_integer and nu.is_nonnegative:
```
Why should `nu` be a nonnegative integer?

0sidharth replied:
For `besselj` - `nu` does not necessarily need to be an integer, further the nonnegative case is handled by the existing code already. I have instead added this as a special case now, as I think `nu` being integer would be used quite often, and having an iterative method would speed that up.

For `bessely` - `nu` needs to be an integer, but not necessarily nonnegative, according to the series expansion formula that I have implemented (now appended in the PR description).

0sidharth replied:
I am not able to figure out how I can add a test for something like - 
```py
assert N(besselj(0.5, x).series(x, n=4), 5) == \                            
        -0.13298*x**2.5 + 0.79788*x**0.5 + O(x**4)
```
It keeps giving assertion error as the coefficients do not equate. Can you help me in fixing this? I have tried a few different functions like `eq`, `S`, `Float` which didn't help

0sidharth replied:
Is something like - `(a - b).subs(x, 2).removeO()` where a is the LHS and b is the RHS alright?

jksuom replied:
I don't think that tests with floats are necessary in this case. They are hard to make right because of rounding errors.

jksuom replied:
The iterative method should work with noninteger values of nu as well (with `gamma(nu + 1)` instead of `factorial(nu)`).

jksuom commented on sympy/functions/special/bessel.py:
```diff
@@ -216,11 +217,44 @@ def _eval_rewrite_as_bessely(self, nu, z, **kwargs):
+
+        if exp.is_positive:
+            newn = ceiling(n/exp)
+            r = z._eval_nseries(x, n, logx, cdir).removeO()
```
```suggestion
            r = (z/2)._eval_nseries(x, n, logx, cdir).removeO()
```
This way 2 and 4 can be deleted from the denominators.

jksuom commented on sympy/functions/special/bessel.py:
```diff
@@ -216,11 +217,44 @@ def _eval_rewrite_as_bessely(self, nu, z, **kwargs):
+            r = z._eval_nseries(x, n, logx, cdir).removeO()
+            t = (z**2)._eval_nseries(x, n, logx, cdir).removeO()
+
+            term = 1/gamma(nu + 1)
```
If `nu` is a negative integer, then this is 0. We can use the equality `besselj(-n, x) = (-1)**n*besselj(n, x)`. So something like this:
```
            if nu.is_integer and nu.is_negative:
                nu = -nu
                term = (-1)**nu/factorial(nu)
            else:
                term = 1/gamma(nu + 1)
```

0sidharth replied:
`nu` cannot be a negative integer. It is handled in `eval` method - https://github.com/sympy/sympy/blob/b564d9ba2705ee2978766e1ee2102750834df68b/sympy/functions/special/bessel.py#L187-L188

jksuom replied:
This does not mean that a symbolic nu were negative. But it will work for numerical values and that will probably suffice.

0sidharth committed:
Optimizations to bessel series expansions

jksuom commented on sympy/functions/special/bessel.py:
```diff
@@ -243,15 +242,16 @@ def _eval_nseries(self, x, n, logx, cdir=0):
             s = [term]
             for k in range(1, (newn + 1)//2):
-                term *= -t/(4*k*(nu + k))
+                term *= -t/(k*(nu + k))
```
Maybe the expand + o  truncation could be applied also here.

jksuom commented on sympy/functions/special/bessel.py:
```diff
@@ -216,11 +216,46 @@ def _eval_rewrite_as_bessely(self, nu, z, **kwargs):
+            r = (z/2)._eval_nseries(x, n, logx, cdir).removeO()
+            t = ((r**2).expand() + o).removeO()
+
+            term = 1/gamma(nu + 1)
```
Could this be `term = r**nu/gamma(nu + 1)`?

0sidharth replied:
Oh, that is a nice change, more terms would be truncated this way. Will make the change.

I am concerned about the case where n is given a small value, say 1, in which case r would be 0, and if `nu` is a negative non-integer, `r**nu` evaluates to `zoo`. In the case of `r = 0`, should I add a condition that would simply return `o`? (`o = O(x**n, x`)
Or would any other value be more appropriate to return?

jksuom replied:
>  that would simply return o

That looks good to me.

jksuom replied:
Also consider this: https://github.com/sympy/sympy/pull/21049#issuecomment-797616965

jksuom commented:
I think the code is now ready to be merged. Can you first squash irrelevant commits?

0sidharth commented:
I squashed and reworded a few commits. Should I squash the commits further?

jksuom commented:
This looks good.Thanks!

I have enabled auto-merge.

0sidharth commented:
Thank you for your extensive review of this PR!

anutosh491 commented:
Hi @jksuom @0sidharth - While working on series for other bessel functions , I realize that the general expansions don't work as expected . Something like this `besselj(nu, x).series(x)` would give `NotImplementedError` . I realized there's something related to the Order term which is responsible for this being caught up !!

I also realized that it has been pointed out before here in the Op
>There is an issue with besselj(n, x).series(x) that returns NotImplementedError: Result depends on the sign of sign(n - 6). This issue is arising due to the Order term which is x^n O(x**6) and I think sympy does not know how to handle this for any integer n <= -1. I can't see how it would be possible to resolve this, so besselj(n, x).series(x) is not supported currently.

Hence I would like to ask if there any way to tackle this ?? Maybe getting this case solved separately ! I would like to have your views regarding the best possible approach here as it is obviously quite nice to have support for general cases not only for bessel but for any special function

anutosh491 commented:
Also another thing I see going wrong here and I plan to fix in coming days is `bessely` missing out on terms for positive order (which would also affect negative terms)

```
>>> bessely(2, x).series(x, 0, 4)     # missing term -1/pi
-4/(pi*x**2) + x**2*(log(x)/(4*pi) - log(2)/(4*pi) - (3/2 - 2*EulerGamma)/(8*pi)) + O(x**4*log(x))
```
![image](https://user-images.githubusercontent.com/87052487/160225346-95dfe79d-b18f-4801-b7ae-ae76b19a3d88.png)

```
>>> bessely(1, x).series(x, 0, 4)    # missing term -2/(pi*x)
x*(log(x)/pi - log(2)/pi - (1 - 2*EulerGamma)/(2*pi)) + x**3*(-log(x)/(8*pi) + (5/2 - 2*EulerGamma)/(16*pi) + log(2)/(8*pi)) + O(x**4*log(x))
```
![image](https://user-images.githubusercontent.com/87052487/160225376-82266321-de7c-4ba7-b85c-164a6dabb518.png)

```
>>> bessely(3, x).series(x, 0, 4)    # missing term -2/(pi*x) and wrongly returned term -4/(pi*x)
-16/(pi*x**3) - 4/(pi*x) + x**3*(log(x)/(24*pi) - log(2)/(24*pi) - (11/6 - 2*EulerGamma)/(48*pi)) + O(x**4*log(x))
```
![image](https://user-images.githubusercontent.com/87052487/160225408-84cdd442-d3e0-4d32-9ea2-b2a36f0871dc.png)

The results attached are from Wolfram and the block responsible for this is this one which needs to be corrected!
```
            if nu > S.One:
                term = r**(-nu)*factorial(nu - 1)/pi
                b.append(term)
                for k in range(1, nu - 1):
                    term *= t*(nu - k - 1)/k
                    term = (_mexpand(term) + o).removeO()
                    b.append(term)
```
